### PR TITLE
Issue #7608: Update doc for DesignForExtension

### PR DESCRIPTION
**Pull request by Sloent**
Issue #7608 : Code examples have been added to the original.

**How the website looks like after I add code example**
图1

图2

图3

图4

**output of default example:**

**Basic example for comparison**
```
C:\Users\SHD\Desktop\checkstyle-master>javac test.java
C:\Users\SHD\Desktop\checkstyle-master>type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension"/>
  </module>
</module>
C:\Users\SHD\Desktop\checkstyle-master>type test.java
 public class A {
    protected void foo1() {};  // OK. Abstract method.

   protected static void foo2() {};  // OK. protected and static method.

   protected final void foo3() {};  // OK. protected and final method.
   }

 public class B extends A {
   protected void foo1() {int arg; };  // Violation. The @Override annotation is missing.

    protected static void foo2() {int arg; };  // Violation. The @Override annotation is missing.

    protected final void foo3() {int arg; };  // Violation. The @Override annotation is missing.
    }

  public class C {
    private int bar;

    public int foo4() { return 1; }  // Violation, required javadoc phrase not in comment.

    private void foo5() {foo6();}  // OK. Private method.

    protected void foo6() { };  // OK. No implementation.

    public abstract void foo7();  // OK. Abstract method.

    /*
    * This
    * implementation ...
    */
    public int foo8() {return 8;}  // OK. Required javadoc phrase in comment.

    /*
    * Do not extend ...
    */
    public int foo9() {return 9;}  // Violation, required javadoc phrase not in comment.

    @Override
    public String toString() {  // Violation. No javadoc for @Override method.
      return "";
    }
  }
C:\Users\SHD\Desktop\checkstyle-master>set RUN_LOCALE="-Duser.language=en -Duser.country=US"

C:\Users\SHD\Desktop\checkstyle-master>java %RUN_LOCALE% -jar checkstyle-10.6.0-all.jar -c config.xml test.java
Starting audit...
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:10:4: Class 'B' looks like designed for extension (can be subclassed), but the method 'foo1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'B' final or making the method 'foo1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:20:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo4' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo4' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:32:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo8' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo8' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:37:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo9' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo9' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:39:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 5 errors.
```

**Example for Config Property igonredAnnotations**
```
C:\Users\SHD\Desktop\checkstyle-master>javac test.java
C:\Users\SHD\Desktop\checkstyle-master>type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension">
      <property name="ignoredAnnotations" value="Override"/>
    </module>
  </module>
</module>
C:\Users\SHD\Desktop\checkstyle-master>type test.java
public class A {
    protected void foo1() {};

   protected static void foo2() {};

   protected final void foo3() {};
   }

 public class B extends A {
   protected void foo1() {int arg; };

    protected static void foo2() {int arg; };

    protected final void foo3() {int arg; };
    }

  public class C {
    private int bar;

    public int foo4() { return 1; }

    private void foo5() {foo6();}

    protected void foo6() { };

    public abstract void foo7();

    /*
    * This
    * implementation ...
    */
    public int foo8() {return 8;}

    /*
    * Do not extend ...
    */
    public int foo9() {return 9;}

    @Override
    public String toString() {
      return "";
    }
  }
C:\Users\SHD\Desktop\checkstyle-master>set RUN_LOCALE="-Duser.language=en -Duser.country=US"

C:\Users\SHD\Desktop\checkstyle-master>java %RUN_LOCALE% -jar checkstyle-10.6.0-all.jar -c config.xml test.java
Starting audit...
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:10:4: Class 'B' looks like designed for extension (can be subclassed), but the method 'foo1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'B' final or making the method 'foo1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:20:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo4' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo4' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:32:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo8' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo8' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:37:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo9' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo9' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 4 errors.
```

**Example for Config Property requiredJavadocPhrase (one line)**
```
C:\Users\SHD\Desktop\checkstyle-master>javac test.java
C:\Users\SHD\Desktop\checkstyle-master>type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension">
      <property name="requiredJavadocPhrase" value="This implementation"/>
    </module>
  </module>
</module>
C:\Users\SHD\Desktop\checkstyle-master>type test.java
public class A {
    protected void foo1() {};

   protected static void foo2() {};

   protected final void foo3() {};
   }

 public class B extends A {
   protected void foo1() {int arg; };

    protected static void foo2() {int arg; };

    protected final void foo3() {int arg; };
    }

  public class C {
    private int bar;

    public int foo4() { return 1; }

    private void foo5() {foo6();}

    protected void foo6() { };

    public abstract void foo7();

    /*
    * This
    * implementation ...
    */
    public int foo8() {return 8;}

    /*
    * Do not extend ...
    */
    public int foo9() {return 9;}

    @Override
    public String toString() {
      return "";
    }
  }
C:\Users\SHD\Desktop\checkstyle-master>set RUN_LOCALE="-Duser.language=en -Duser.country=US"

C:\Users\SHD\Desktop\checkstyle-master>java %RUN_LOCALE% -jar checkstyle-10.6.0-all.jar -c config.xml test.java
Starting audit...
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:10:4: Class 'B' looks like designed for extension (can be subclassed), but the method 'foo1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'B' final or making the method 'foo1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:20:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo4' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo4' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:32:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo8' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo8' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:37:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo9' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo9' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:39:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 5 errors.
```

**Example for Config Property requiredJavadocPhrase (span in multiple line)**
```
C:\Users\SHD\Desktop\checkstyle-master>javac test.java
C:\Users\SHD\Desktop\checkstyle-master>type config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="DesignForExtension">
      <property name="requiredJavadocPhrase" value="This[\s\S]*implementation"/>
    </module>
  </module>
</module>
C:\Users\SHD\Desktop\checkstyle-master>type test.java
public class A {
    protected void foo1() {};

   protected static void foo2() {};

   protected final void foo3() {};
   }

 public class B extends A {
   protected void foo1() {int arg; };

    protected static void foo2() {int arg; };

    protected final void foo3() {int arg; };
    }

  public class C {
    private int bar;

    public int foo4() { return 1; }

    private void foo5() {foo6();}

    protected void foo6() { };

    public abstract void foo7();

    /*
    * This
    * implementation ...
    */
    public int foo8() {return 8;}

    /*
    * Do not extend ...
    */
    public int foo9() {return 9;}

    @Override
    public String toString() {
      return "";
    }
  }
C:\Users\SHD\Desktop\checkstyle-master>set RUN_LOCALE="-Duser.language=en -Duser.country=US"

C:\Users\SHD\Desktop\checkstyle-master>java %RUN_LOCALE% -jar checkstyle-10.6.0-all.jar -c config.xml test.java
Starting audit...
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:10:4: Class 'B' looks like designed for extension (can be subclassed), but the method 'foo1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'B' final or making the method 'foo1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:20:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo4' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo4' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:32:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo8' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo8' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:37:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo9' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo9' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] C:\Users\SHD\Desktop\checkstyle-master\test.java:39:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 5 errors.
```

### CODE REVIEW

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Please make CI happy.

Item:

**Comment by nrmancuso** [comment]
Github, generate site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java [hunk 1]
```diff
@@ -383,7 +415,7 @@ public class DesignForExtensionCheck extends AbstractCheck {
      * Specify annotations which allow the check to skip the method from validation.
      */
     private Set<String> ignoredAnnotations = Arrays.stream(new String[] {"Test", "Before", "After",
-        "BeforeClass", "AfterClass", }).collect(Collectors.toSet());
+            "BeforeClass", "AfterClass", }).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (Sloent) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java:418]: This is my first time contributing to an open source project. I will try to make CI happy and restore code/format changes outside the scope of the problem.

**Comment by nrmancuso** [comment]
Attention @romani @rnveach @Vyom-Yadav  - if site generation seems to be failing on first time contributor's PR, it is due to requirement that actions are approved by a maintainer

**Comment by Sloent** [comment]
The Job test and OpenJDK veryfy that failed to perform a series of tests and prompt the ① An error occurred while "DesignForExtensionCheck.java". This error occurs when running a JUnit test named "XdocsJavaDocsTest" that attempts to check JavaDoc comments in the Checkstyle source code. Unfortunately, the error message does not provide enough information to determine the root cause of the problem; ② Because the Surefire plugin failed to perform tests.

Then I looked at the log output from Surefire and, based on the test results, Test class ` com.puppycrawl.tools.checkstyle.checks.design.DesignForExtensionCheckTest ` methods in ` testIgnoredAnnotationsOption ` A failure was reported.  This test uses the 'DesignForExtensionCheck' checker in the Checkstyle plug-in to check whether the class is designed to be inheritable.  Test expectation in ` InputDesignForExtensionIgnoredAnnotations.java ` file line 176 and found a do not conform to the requirements of the lack of method of javadoc comments ` dontUse4 `.  But what was actually found at the line was a Chinese error message indicating that javadoc should be added for inheritable classes and that javadoc should be added for the 'dontUse4' method.  Therefore, there is no way to determine which specific file the 'dontUse4' method is in based on the test results.

However, I in addition to DesignForExtensionCheck.java and config_design.xml is changed, other files are not changed, but they seem to lead to an error in the CI to check.

This is the summary of my attempt to eliminate mistakes today. I sincerely hope that the tutor can give me advice!

**Comment by romani** [comment]
Double quote symbols are different
Compare
https://github.com/checkstyle/checkstyle/blob/bf9b7b2cd8599df78d183244b437875c59c4e21c/src/xdocs/config_design.xml#L51
Vs
https://github.com/checkstyle/checkstyle/blob/bf9b7b2cd8599df78d183244b437875c59c4e21c/src/xdocs/config_design.xml#L59

Line 51 is correct , line 59 some Unicode.

**Comment by romani** [comment]
GitHub, generate website

**Comment by Sloent** [comment]
There are no more "" errors in the config_design.xml file, but CI checking is still not promising. I have no idea how to modify to make CI check pass.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java [hunk 2]
```diff
@@ -415,7 +415,7 @@ public class DesignForExtensionCheck extends AbstractCheck {
      * Specify annotations which allow the check to skip the method from validation.
      */
     private Set<String> ignoredAnnotations = Arrays.stream(new String[] {"Test", "Before", "After",
-            "BeforeClass", "AfterClass", }).collect(Collectors.toSet());
+        "BeforeClass", "AfterClass", }).collect(Collectors.toSet());
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/design/DesignForExtensionCheck.java]: revert this fix, it is not required.

**Review by romani — CHANGES_REQUESTED** [review]
item:

**Comment by romani** [comment]
I do not see reason in this update.
you examples are wrong,  lets close this PR and move on to next issues.

```
$ cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    
    <module name="DesignForExtension"/>
  </module>
</module>
rivanov@p5510:/var/tmp$ cat Test.java
public class A {
    protected void foo1() {};  // OK. Abstract method.

    protected static void foo2() {};  // OK. protected and static method.

    protected final void foo3() {};  // OK. protected and final method.
}

public class B extends A {
    protected void foo1() { int arg; };  // Violation. The @Override annotation is missing.

    protected static void foo2() { int arg; };  // Violation. The @Override annotation is missing.

    protected final void foo3() { int arg; };  // Violation. The @Override annotation is missing.
}

public class C {
    private int bar;

    public int foo4() { return 1; }  // Violation, required javadoc phrase not in comment.

    private void foo5() { foo6(); }  // OK. Private method.

    protected void foo6() { }  // OK. No implementation.

    public abstract void foo7();  // OK. Abstract method.

    /**
    * This
    * implementation ...
    */
    public int foo8() {return 8;}  // OK. Required javadoc phrase in comment.

    /**
    * Do not extend ...
    */
    public int foo9() {return 9;}  // Violation, required javadoc phrase not in comment.

    @Override
    public String toString() {  // Violation. No javadoc for @Override method.
    return "";
    }
}
rivanov@p5510:/var/tmp$ cat -n Test.java
     1	public class A {
     2	    protected void foo1() {};  // OK. Abstract method.
     3	
     4	    protected static void foo2() {};  // OK. protected and static method.
     5	
     6	    protected final void foo3() {};  // OK. protected and final method.
     7	}
     8	
     9	public class B extends A {
    10	    protected void foo1() { int arg; };  // Violation. The @Override annotation is missing.
    11	
    12	    protected static void foo2() { int arg; };  // Violation. The @Override annotation is missing.
    13	
    14	    protected final void foo3() { int arg; };  // Violation. The @Override annotation is missing.
    15	}
    16	
    17	public class C {
    18	    private int bar;
    19	
    20	    public int foo4() { return 1; }  // Violation, required javadoc phrase not in comment.
    21	
    22	    private void foo5() { foo6(); }  // OK. Private method.
    23	
    24	    protected void foo6() { }  // OK. No implementation.
    25	
    26	    public abstract void foo7();  // OK. Abstract method.
    27	
    28	    /**
    29	    * This
    30	    * implementation ...
    31	    */
    32	    public int foo8() {return 8;}  // OK. Required javadoc phrase in comment.
    33	
    34	    /**
    35	    * Do not extend ...
    36	    */
    37	    public int foo9() {return 9;}  // Violation, required javadoc phrase not in comment.
    38	
    39	    @Override
    40	    public String toString() {  // Violation. No javadoc for @Override method.
    41	    return "";
    42	    }
    43	}
rivanov@p5510:/var/tmp$ java -jar checkstyle-10.9.2-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] /var/tmp/Test.java:10:5: Class 'B' looks like designed for extension (can be subclassed), but the method 'foo1' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'B' final or making the method 'foo1' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /var/tmp/Test.java:20:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'foo4' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'foo4' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
[ERROR] /var/tmp/Test.java:39:5: Class 'C' looks like designed for extension (can be subclassed), but the method 'toString' does not have javadoc that explains how to do that safely. If class is not designed for extension consider making the class 'C' final or making the method 'toString' static/final/abstract/empty, or adding allowed annotation for the method. [DesignForExtension]
Audit done.
Checkstyle ends with 3 errors.

```

**Pull request closed without merging**